edge-modules/metrics-collector/src/ModuleClientWrapper/BasicModuleClientWrapper.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Edge.Util;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Devices.Edge.Azure.Monitor.ModuleClientWrapper
+{
+    public class BasicModuleClientWrapper : IDisposable, IModuleClientWrapper
+    {
+        ModuleClient inner;
+        SemaphoreSlim moduleClientLock;
+
+        public BasicModuleClientWrapper(ModuleClient moduleClient, SemaphoreSlim moduleClientLock)
+        {
+            this.inner = moduleClient;
+            this.moduleClientLock = Preconditions.CheckNotNull(moduleClientLock);
+        }
+
+        public static async Task<BasicModuleClientWrapper> BuildModuleClientWrapperAsync()
+        {
+            SemaphoreSlim moduleClientLock = new SemaphoreSlim(1, 1);
+            ModuleClient moduleClient = await InitializeModuleClientAsync();
+            return new BasicModuleClientWrapper(moduleClient, moduleClientLock);
+        }
+
+        public async Task RecreateClientAsync()
+        {
+
+            try
+            {
+                await this.moduleClientLock.WaitAsync();
```

Comment:
> Why add it here? If WaitAsync throws, is there a scenario where this code will have the lock (and therefore need to catch the exception to release the lock)? I would assume not.

I think if we throw we need to still make sure to release the lock. While I was testing I noticed a situation where execution hung caused by not releasing lock in either case.

Reply:
SemaphoreSlim.WaitAsync() will only throw an exception (per the documentation) if it has already been disposed. If that happens, and we catch the exception and call SemaphoreSlim.Release(), it will throw another exception. So I think the WaitAsync() call should be outside the try/catch block. If WaitAsync() throws, that's a violation of an invariant (it means the semaphore was prematurely disposed--a bug) and we want to fail fast.

Reply:
I understand what you mean. I think it is cleaner if we wait on the lock outside the try based on your logic. Will fix.